### PR TITLE
enable provider/extension side copy to clipboard support

### LINKED ISSUE #10123: Azure Data Studio doesn't work reliably when attempting to copy large amounts of data
Issue Type: <b>Bug</b>

Perform a query with roughly 15K results.  Attempt to copy the results to clipboard and Azure Data Studio will hang and eventually crash.

Azure Data Studio version: azuredatastudio 1.16.1 (16801c522df68e63218b39ccb5b4a7431f9bf667, 2020-03-23T22:08:06.892Z)
OS version: Windows_NT x64 10.0.19041

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-10710U CPU @ 1.10GHz (12 x 1608)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>protected_video_decode: enabled<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>viz_display_compositor: enabled_on<br>viz_hit_test_surface_layer: disabled_off_ok<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|31.84GB (12.37GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details>

### PR DESCRIPTION
**Pull request by alanrenmsft**
This PR fixes #10123

1. Use the extension/provider's copy to clipboard implementation if available to optimize the performance.
2. Add a setting to force to use the ADS side copy to clipboard implementation in case the provider implementation doesn't work on certain Linux platforms.

1. Provider's own copy to clipboard implementation.
provider-copy

2. ADS's copy to clipboard implementation.
ads-copy

3. copy small selections
copy-small-set

STS changes

image

### CODE REVIEW

**Comment by Charles-Gagnon** [comment]
Can you include a gif of what it looks like when only copying a small set of values? (e.g. <10 rows with <5 columns)

**Comment by alanrenmsft** [comment]
> Can you include a gif of what it looks like when only copying a small set of values? (e.g. <10 rows with <5 columns)

@Charles-Gagnon  done

File: src/sql/workbench/services/query/common/queryRunner.ts [hunk 1]
```diff
@@ -559,14 +574,27 @@ export class QueryGridDataProvider implements IGridDataProvider {
 		return this.copyResultsAsync(selection, includeHeaders, tableView);
 	}
 
-	private async copyResultsAsync(selection: Slick.Range[], includeHeaders?: boolean, tableView?: IDisposableDataProvider<Slick.SlickData>): Promise<void> {
+	private async copyResultsAsync(selections: Slick.Range[], includeHeaders?: boolean, tableView?: IDisposableDataProvider<Slick.SlickData>): Promise<void> {
 		try {
-			await copySelectionToClipboard(this._clipboardService, this._notificationService, this, selection, includeHeaders, tableView);
+			const providerId = this.queryRunner.getProviderId();
+			const providerSupportCopyResults = this._capabilitiesService.getCapabilities(providerId).connection.supportCopyResultsToClipboard;
```
Comment (kburtram) [inline src/sql/workbench/services/query/common/queryRunner.ts:580]: where do we detect if we're running the extension host in "remote mode" so we don't use the provider copy in that case?

**Review by kburtram — APPROVED** [review]
LGTM though I didn't see where we disable the provider copy handler in remote mode?  I suppose that's only important when we backport the improvement to vscode though.

**Review by aasimkhan30 — APPROVED** [review]

**Review by brian-harris — APPROVED** [review]

File: src/sql/workbench/contrib/query/browser/query.contribution.ts [hunk 2]
```diff
@@ -404,6 +404,11 @@ const queryEditorConfiguration: IConfigurationNode = {
 			'description': localize('queryEditor.results.copyRemoveNewLine', "Configuration options for copying multi-line results from the Results View"),
 			'default': true
 		},
+		'queryEditor.results.preferProvidersCopyHandler': {
+			'type': 'boolean',
+			'description': localize('queryEditor.results.preferProvidersCopyHandler', "Whether the copy result request should be handled by the query provider when it is supported. The default value is true, set this to false in case the provider's copy request handler doesn't work on certain platforms."),
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/query/browser/query.contribution.ts:409]: Not sure having a reason here is necessary, there are other reasons why people might want to disable this so better to keep it generic IMO.

```suggestion
			'description': localize('queryEditor.results.preferProvidersCopyHandler', "Whether the copy result request should be handled by the query provider when it is supported. The default value is true, set this to false to force all copy handling to be done by Azure Data Studio."),
```

**Review by Charles-Gagnon — APPROVED** [review]

File: src/sql/workbench/contrib/query/browser/query.contribution.ts (hunk 2 above)
Comment (alanrenmsft) [reply src/sql/workbench/contrib/query/browser/query.contribution.ts:409]: done.

File: src/sql/platform/capabilities/common/capabilitiesService.ts [hunk 3]
```diff
@@ -108,6 +108,13 @@ export interface ConnectionProviderProperties {
 	 * Connection string options for the connection provider
 	 */
 	connectionStringOptions?: ConnectionStringOptions;
+
+	/**
+	 * Indicates whether the provider support copy results to clipboard. Default value is false.
+	 * If true, the copy results to clipboard will be delegated to the provider to avoid passing large amount of data using the RPC channel.
+	 * Otherwise ADS will handle the copy request on the UI side.
+	 */
+	supportCopyResultsToClipboard?: boolean;
```
[the remark this answers is not in the document]
Comment (alanrenmsft) [reply src/sql/platform/capabilities/common/capabilitiesService.ts:117]: done

File: src/sql/workbench/services/query/common/queryRunner.ts (hunk 1 above)
Comment (alanrenmsft) [reply src/sql/workbench/services/query/common/queryRunner.ts:580]: that will be in the vscode-mssql extension.

**Pull request merged**